### PR TITLE
Codependence by Marti / Codependence toctree / Bug fixes

### PR DESCRIPTION
**Pull request by PanPip**
# Description
**Codependence by Marti:**

This is the implementation of Spearman's Rho, the GPR, and the GNPF codependency metrics described in the work of Gautier Marti "Some contributions to the clustering of financial time series and applications to credit default swaps".

The code is built in the form of separate functions.

The unit test and Sphinx documentation are also provided with the code.

**Codependence toctree:**
This PR also creates a separate toctree for Codependence metrics, updates the documentation, docstrings, and some of the functions in that module.

**Bug fixes:**
Currently this PR fixes: #355, #262, #362 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Unit tests of the Codependence module were updated with tests for the new functions.

- [x] test_codependence.py

**Test Configuration**:
* Windows 10
* PyCharm

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/source/implementations/codependence.rst [hunk 1]
```diff
@@ -324,14 +324,16 @@ Codependence Matrix
 This module consists two functions that generate the following:
 
 1. **Dependence Matrix** to compute dependence of a given matrix using various codependence method like Mutual Information,
```
Comment (aditya1702) [inline docs/source/implementations/codependence.rst:326]: "method" -> "methods"

File: docs/source/implementations/codependence_marti.rst [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+   The following implementations and documentation closely follows the work of Gautier Marti:
+   `Some contributions to the clustering of financial time series and applications to credit default swaps <https://www.researchgate.net/publication/322714557>`_.
+
+=====================
+Codependence by Marti
+=====================
+
+The work mentioned above introduces a new approach of representing the random variables that splits apart dependency and
+distribution without losing any information. It also contains a distance metric between two financial time series based
+on the novel approach.
```
Comment (aditya1702) [inline docs/source/implementations/codependence_marti.rst:13]: "the" -> "a"

Comment (PanPip) [reply docs/source/implementations/codependence_marti.rst]: Thank you for the review 🙂

File: docs/source/codependence/information_theory_metrics.rst [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+This optimal binning method is used in the mutual information and variation of information functions.
+
+Implementation
+##############
+
+.. autofunction:: get_optimal_number_of_bins
+
+
+Examples
```
Comment (aditya1702) [inline docs/source/codependence/information_theory_metrics.rst:147]: Can you change this to "Example Code", just to be more clear for the users

File: mlfinlab/codependence/codependence_matrix.py [hunk 4]
```diff
@@ -70,12 +75,14 @@ def get_dependence_matrix(df: pd.DataFrame, dependence_method: str) -> pd.DataFr
...
-    Apply distance operator to a dependence matrix.
+    Applies distance operator to a dependence matrix.
+
+    This allows to turn a correlation matrix into a distance matrix. Distances used are true metrics.
 
     :param X: (pd.DataFrame) Dataframe to which distance operator to be applied.
     :param distance_metric: (str) The distance operator to be used for generating the distance matrix.
-       The methods that can be applied are: 'angular', 'squared_angular' and 'absolute_angular'.
-    :return: (pd.DataFrame) Distance matrix
+                                  Supported methods are: 'angular', 'squared_angular' and 'absolute_angular'.
```
Comment (aditya1702) [inline mlfinlab/codependence/codependence_matrix.py:84]: You can change these to - Supported methods are: ``angular``, ``squared_angular`` etc... This just shows up more nicely on read-the-docs. I have done the same for portfolio optimisation docs. You can give it a look to see how it is rendered

File: docs/source/codependence/information_theory_metrics.rst (hunk 3 above)
Comment (PanPip) [reply docs/source/codependence/information_theory_metrics.rst:147]: I was just following the style that we use in the PortOpt and other modules. 

I think I'll leave it like this for now and then make a PR to change the names of Code examples globally in MlFinLab. 

Thank you for noticing!

File: mlfinlab/codependence/codependence_matrix.py (hunk 4 above)
Comment (PanPip) [reply mlfinlab/codependence/codependence_matrix.py:84]: Thanks. I'll make this change. 🙂

**Pull request merged**